File: 0139-word-break/0139-word-break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:

        #Time Complexity : O(N^2)
        #Space Complexity : O(M) + O(N)        

        word_set = set(wordDict) #O(M)
        dp = [False]*(len(s)+1) #O(N)
        dp[0] = True

        for i in range(1, len(s)+1): #O(N)
            for j in range(i): #O(N)
                if dp[j] and s[j:i] in word_set:
                    dp[i] = True
                    break

        return dp[len(s)]

File: 0139-word-break/0139-word-break.java
class Solution {

    public boolean wordBreak(String s, List<String> wordDict) {

        int n = s.length();
        boolean[] isExist = new boolean[n+1];
        isExist[0] = true;

        for(int i=1; i<=n; i++){
            for(String word : wordDict){
                if(word.length() <= i && isExist[i-word.length()]){
                    if(word.equals(s.substring(i-word.length(), i))){
                        isExist[i] = true;
                    }
                }
            }            
        }
        return isExist[n];
    }
}